### PR TITLE
Use single user

### PR DESCRIPTION
**Pull request by DomGarguilo**
Supersedes #58

### CODE REVIEW

File: src/main/java/org/apache/accumulo/proxy/ProxyServer.java [hunk 1]
```diff
@@ -129,8 +125,10 @@ public class ProxyServer implements AccumuloProxy.Iface {
   public static final Logger logger = LoggerFactory.getLogger(ProxyServer.class);
   public static final String RPC_ACCUMULO_PRINCIPAL_MISMATCH_MSG = "RPC"
       + " principal did not match requested Accumulo principal";
-  @SuppressWarnings("deprecation")
-  protected org.apache.accumulo.core.client.Instance instance;
+
+  protected AccumuloClient client;
+
+  private final String secret;
```
Comment (DomGarguilo) [inline src/main/java/org/apache/accumulo/proxy/ProxyServer.java:131]: This is one of the main changes in this PR. The addition of this shared secret allows for the client to authenticate to the proxy service without telling the proxy service how to authenticate to accumulo.

File: src/test/java/org/apache/accumulo/proxy/ProxyServerTest.java [hunk 2]
```diff
@@ -69,14 +68,14 @@ public void updateAndFlushClosesWriterOnExceptionFromAddCells() throws Exception
...
-        .addMockedMethod("getWriter", ByteBuffer.class, String.class, WriterOptions.class)
+        .addMockedMethod("getWriter", String.class, String.class, WriterOptions.class)
         .addMockedMethod("addCellsToWriter", Map.class, BatchWriterPlusProblem.class).createMock();
     BatchWriter writer = EasyMock.createMock(BatchWriter.class);
     BatchWriterPlusProblem bwpe = new BatchWriterPlusProblem();
     bwpe.writer = writer;
     MutationsRejectedException mre = EasyMock.createMock(MutationsRejectedException.class);
 
-    final ByteBuffer login = ByteBuffer.wrap("my_login".getBytes(UTF_8));
+    final String login = "my_login";
```
Comment (ctubbsii) [inline src/test/java/org/apache/accumulo/proxy/ProxyServerTest.java:78]: The change to use a shared proxy secret instead of the previous login mechanism is okay, but it left a bunch of tests with confusing variable names, which make the tests hard to follow.

File: src/test/java/org/apache/accumulo/proxy/its/ProxyDurabilityIT.java [hunk 3]
```diff
@@ -98,6 +98,10 @@ public void testDurability() throws Exception {
       proxyProps.put("tokenClass", PasswordToken.class.getName());
       proxyProps.putAll(getClientProperties());
 
+      String login = ROOT_PASSWORD;
+
+      proxyProps.put("secret", login);
+
```
Comment (ctubbsii) [inline src/test/java/org/apache/accumulo/proxy/its/ProxyDurabilityIT.java:104]: "secret" and "login" are significantly overused as variables, constants, etc. It's very hard to follow the tests, when the root password is being stored in a variable called "login" that is actually the shared secret for authenticating to the proxy. There needs to be greater clarity between all of these, with better variable names, and better test values, and less reuse between them... I suspect some tests "just work" accidentally, because we used "secret" more than once, so even though we're probably passing the wrong variable, it still passes, because both values are "secret". Everything should have unique values.

File: src/test/java/org/apache/accumulo/proxy/its/SimpleProxyBase.java [hunk 4]
```diff
@@ -162,13 +163,12 @@ protected Duration defaultTimeout() {
...
   private org.apache.accumulo.proxy.thrift.AccumuloProxy.Client client;
 
-  private static final Map<String,String> properties = new HashMap<>();
   private static String hostname;
   private static String proxyPrimary;
   private static String clientPrincipal;
   private static File clientKeytab;
 
-  private ByteBuffer creds = null;
+  private static String creds;
```
Comment (ctubbsii) [inline src/test/java/org/apache/accumulo/proxy/its/SimpleProxyBase.java:171]: `creds` is another one whose name is overused and needs more clarity. In this case, it's storing the shared secret for the proxy.

File: src/test/java/org/apache/accumulo/proxy/its/SimpleProxyBase.java [hunk 5]
```diff
@@ -1232,8 +1224,8 @@ public void testSiteConfiguration() throws Exception {
     // get something we know is in the site config
     MiniAccumuloClusterImpl cluster = SharedMiniClusterBase.getCluster();
     Map<String,String> cfg = client.getSiteConfiguration(creds);
-    assertTrue(cfg.get("instance.dfs.dir")
-        .startsWith(cluster.getConfig().getAccumuloDir().getAbsolutePath()));
+    assertEquals(new File(new URI(cfg.get("instance.volumes"))).getCanonicalPath(),
+        cluster.getConfig().getAccumuloDir().getCanonicalPath());
```
Comment (ctubbsii) [inline src/test/java/org/apache/accumulo/proxy/its/SimpleProxyBase.java:1228]: I fixed this test. It should be good now.

File: src/test/java/org/apache/accumulo/proxy/its/SimpleProxyBase.java [hunk 6]
```diff
@@ -1428,7 +1420,7 @@ public void userAuthentication() throws Exception {
       // check password
       assertTrue(
           client.authenticateUser(creds, "root", s2pp(SharedMiniClusterBase.getRootPassword())));
-      assertFalse(client.authenticateUser(creds, "root", s2pp("")));
+      assertFalse(client.authenticateUser(creds, "otheruser", s2pp("")));
```
Comment (ctubbsii) [inline src/test/java/org/apache/accumulo/proxy/its/SimpleProxyBase.java:1423]: This test was broken, but it's not clear why. This might actually require further investigation in Accumulo itself. It seemed to return true, regardless of the password, if the root user tried to authenticate themselves. A good test here might be to create a new user, and try to authenticate that new user, with their new password and a bad password, instead of trying to test with the root user, which is also the current user.

File: src/test/java/org/apache/accumulo/proxy/its/SimpleProxyBase.java [hunk 7]
```diff
@@ -1463,34 +1455,15 @@ public void userManagement() throws Exception {
       client.changeLocalUserPassword(creds, user, password);
       assertTrue(client.authenticateUser(creds, user, s2pp(ByteBufferUtil.toString(password))));
     }
-
-    if (isKerberosEnabled()) {
```
Comment (ctubbsii) [inline src/test/java/org/apache/accumulo/proxy/its/SimpleProxyBase.java:1467]: I'm not really sure we need any of these kerberos checks. If the AccumuloClient is correctly created from props, that client should transparently handle whether it uses a PasswordToken or a KerberosToken in the user's client properties file.

There are a lot of places in the proxy where we have special handling for kerberos... especially for the tests. I'm not actually sure if the proxy supports, or ever supported, running as a kerberos-enabled service itself. If that's something we want to support, it will probably require substantial follow-on work. For now, it might just be useful to know whether the old proxy did or did not support that.

File: src/test/java/org/apache/accumulo/proxy/its/SimpleProxyBase.java [hunk 8]
```diff
@@ -1463,34 +1455,15 @@ public void userManagement() throws Exception {
...
-          otherProxyClient.close();
-        }
-      }
-    } else {
-      // check login with new password
-      client.login(user, s2pp(ByteBufferUtil.toString(password)));
-    }
   }
 
+  @Disabled("This test needs to be reworked, because it tries to switch users")
```
Comment (ctubbsii) [inline src/test/java/org/apache/accumulo/proxy/its/SimpleProxyBase.java:1460]: The worst tests, and the ones that need the most work, are the ones that try to create and switch users. I disabled three of those.

File: src/test/java/org/apache/accumulo/proxy/its/SimpleProxyBase.java [hunk 9]
```diff
@@ -2640,9 +2611,7 @@ private ByteBuffer s2bb(String cf) {
   }
 
   private Map<String,String> s2pp(String cf) {
-    Map<String,String> toRet = new TreeMap<>();
-    toRet.put("password", cf);
-    return toRet;
+    return Map.of("password", cf);
```
Comment (ctubbsii) [inline src/test/java/org/apache/accumulo/proxy/its/SimpleProxyBase.java:2614]: This was a trivial thing I changed. I have no idea what the `s2pp` method names were supposed to be... maybe "string to password property"?

**Review by ctubbsii — COMMENTED** [review]
I think there's been good progress, but the proxy still needs a lot of work. This PR could probably be merged soon, if not now, as an incremental improvement.
